Source/render.cpp

```diff
@@ -422,306 +539,834 @@ inline void DoRenderLine(BYTE *dst, BYTE *src, size_t n, BYTE *tbl, DWORD mask)
+		else
+			break;
+		src += width;
+	}
+}
+
+void RenderLeftTriangle(BYTE *dst, int dstPitch, BYTE *src, const DWORD *mask, const BYTE *tbl, Clip clip)
+{
+	if (clip.width == Width) {
+		if (clip.height == TriangleHeight) {
+			RenderLeftTriangleFull(dst, dstPitch, src, mask, tbl);
+		} else {
+			RenderLeftTriangleClipVertical(dst, dstPitch, src, mask, tbl, clip);
 		}
-		if (arch_draw_type == 1) {
-			mask = &LeftFoliageMask[TILE_HEIGHT - 1];
+	} else if (clip.right == 0) {
+		RenderLeftTriangleClipLeftAndVertical(dst, dstPitch, src, mask, tbl, clip);
+	} else {
+		RenderLeftTriangleClipRightAndVertical(dst, dstPitch, src, mask, tbl, clip);
+	}
+}
+
+void RenderRightTriangleFull(BYTE *dst, int dstPitch, BYTE *src, const DWORD *mask, const BYTE *tbl)
+{
+	for (auto i = 1; i <= LowerHeight; ++i, dst -= dstPitch, --mask) {
+		const auto width = XStep * i;
+		RenderLine(dst, src, width, tbl, *mask);
+		src += width + 2 * (i % 2);
+	}
+	for (auto i = 1; i <= TriangleUpperHeight; ++i, dst -= dstPitch, --mask) {
+		const auto width = Width - XStep * i;
+		RenderLine(dst, src, width, tbl, *mask);
+		src += width + 2 * (i % 2);
+	}
+}
+
+void RenderRightTriangleClipVertical(BYTE *dst, int dstPitch, BYTE *src, const DWORD *mask, const BYTE *tbl, Clip clip)
+{
+	const auto clipY = CalculateDiamondClipY(clip);
+	src += CalculateTriangleSourceSkipLowerBottom(clipY.lowerBottom);
+	const auto lowerMax = LowerHeight - clipY.lowerTop;
+	for (auto i = 1 + clipY.lowerBottom; i <= lowerMax; ++i, dst -= dstPitch, --mask) {
+		const auto width = XStep * i;
+		RenderLine(dst, src, width, tbl, *mask);
+		src += width + 2 * (i % 2);
+	}
+	src += CalculateTriangleSourceSkipUpperBottom(clipY.upperBottom);
+	const auto upperMax = TriangleUpperHeight - clipY.upperTop;
+	for (auto i = 1 + clipY.upperBottom; i <= upperMax; ++i, dst -= dstPitch, --mask) {
+		const auto width = Width - XStep * i;
+		RenderLine(dst, src, width, tbl, *mask);
+		src += width + 2 * (i % 2);
+	}
+}
+
+void RenderRightTriangleClipLeftAndVertical(BYTE *dst, int dstPitch, BYTE *src, const DWORD *mask, const BYTE *tbl, Clip clip)
+{
+	const auto clipY = CalculateDiamondClipY(clip);
+	const auto clipLeft = clip.left;
+	src += CalculateTriangleSourceSkipLowerBottom(clipY.lowerBottom);
+	const auto lowerMax = LowerHeight - clipY.lowerTop;
+	for (auto i = 1 + clipY.lowerBottom; i <= lowerMax; ++i, dst -= dstPitch, --mask) {
+		const auto width = XStep * i;
+		if (width > clipLeft)
+			RenderLine(dst, src + clipLeft, width - clipLeft, tbl, (*mask) << clipLeft);
+		src += width + 2 * (i % 2);
+	}
+	src += CalculateTriangleSourceSkipUpperBottom(clipY.upperBottom);
+	const auto upperMax = TriangleUpperHeight - clipY.upperTop;
+	for (auto i = 1 + clipY.upperBottom; i <= upperMax; ++i, dst -= dstPitch, --mask) {
+		const auto width = Width - XStep * i;
+		if (width > clipLeft)
+			RenderLine(dst, src + clipLeft, width - clipLeft, tbl, (*mask) << clipLeft);
+		else
+			break;

```

Comment:
This feels missleading
```suggestion
		if (width <= clipLeft)
			break;

		RenderLine(dst, src + clipLeft, width - clipLeft, tbl, (*mask) << clipLeft);
```

Reply:
Done